presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -252,6 +260,96 @@ public HiveTableHandle getTableHandle(ConnectorSession session, SchemaTableName
         return new HiveTableHandle(tableName.getSchemaName(), tableName.getTableName());
     }
 
+    private static final String PARTITIONS_TABLE_SUFFIX = "$partitions";
+
+    @Override
+    public Optional<SystemTable> getSystemTable(ConnectorSession session, SchemaTableName tableName)
+    {
+        if (tableName.getTableName().endsWith(PARTITIONS_TABLE_SUFFIX)) {
+            return getPartitionsSystemTable(session, tableName);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<SystemTable> getPartitionsSystemTable(ConnectorSession session, SchemaTableName tableName)
+    {
+        SchemaTableName sourceTableName = new SchemaTableName(
+                tableName.getSchemaName(),
+                tableName.getTableName().substring(0, tableName.getTableName().length() - PARTITIONS_TABLE_SUFFIX.length()));
+
+        HiveTableHandle sourceTableHandle = getTableHandle(session, sourceTableName);
+
+        if (sourceTableHandle == null) {
+            return Optional.empty();
+        }
+
+        List<HiveColumnHandle> partitionColumns = getPartitionColumns(sourceTableName);
+        if (partitionColumns.isEmpty()) {
+            return Optional.empty();
+        }
+
+        List<Type> partitionColumnTypes = partitionColumns.stream()
+                .map(HiveColumnHandle::getTypeSignature)
+                .map(typeManager::getType)
+                .collect(toImmutableList());
+
+        List<ColumnMetadata> partitionSystemTableColumns = partitionColumns.stream()
+                .map(col -> new ColumnMetadata(
+                        col.getName(),
+                        typeManager.getType(col.getTypeSignature()),
+                        col.getComment().orElse(""),
```

Comment:
I think this should be `orElse(null)`. That's what we do in `columnMetadataGetter()`